sky/provision/aws/utils.py

```diff
@@ -0,0 +1,140 @@
+"""Utils for AWS provisioner."""
+import threading
+from typing import Any, Dict
+
+import boto3
+from botocore import config
+import colorama
+import urllib3
+
+from sky import sky_logging
+
+logger = sky_logging.init_logger(__name__)
+
+BOTO_MAX_RETRIES = 12
+
+# ======================== Thread-safe ========================
+# https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html#multithreading-or-multiprocessing-with-resources
+
+_lock = threading.RLock()
+
+
+def create_resource(resource: str,
+                    region: str,
+                    max_attempts: int = BOTO_MAX_RETRIES,
+                    **credentials) -> Any:
+    """Create an AWS resource."""
+    # overhead: 6.69 ms ± 41.5 µs
+    with _lock:
+        # We should prevent caching the resource for thread safety.
+        # Instead, the logic that uses the resource should reuse the same
+        # resource as much as possible.
+        return boto3.resource(
+            resource,
+            region,
+            config=config.Config(retries={'max_attempts': max_attempts}),
+            **credentials)
```

Comment:
Should we directly use the `sky.adaptors.aws.resource()`? There seems to have more logic for handling this thread safety.